conans/test/integration/graph/core/test_build_requires.py

```diff
@@ -585,3 +579,44 @@ def build_requirements(self):
 
         self._check_node(app, "app/0.1@", deps=[tool], dependents=[])
         self._check_node(tool, "gtest/0.1#123", deps=[], dependents=[app])
+
+
+class TestLoops(GraphManagerTest):
+
+    def test_direct_loop_error(self):
+        # app -(br)-> cmake/0.1 -(br itself)-> cmake/0.1....
+        # causing an infinite loop
+        self._cache_recipe("cmake/0.1", GenConanfile().with_tool_requires("cmake/0.1"))
+        deps_graph = self.build_graph(GenConanfile("app", "0.1").with_build_requires("cmake/0.1"),
+                                      install=False)
+
+        assert deps_graph.error.kind == GraphError.LOOP
+
+        # Build requires always apply to the consumer
+        self.assertEqual(2, len(deps_graph.nodes))
+        app = deps_graph.root
+        tool = app.dependencies[0].dst
+
+        self._check_node(app, "app/0.1", deps=[tool], dependents=[])
+        self._check_node(tool, "cmake/0.1#123", deps=[], dependents=[app])
+
+    def test_indirect_loop_error(self):
+        # app -(br)-> gtest/0.1 -(br)-> cmake/0.1 -(br)->gtest/0.1 ....
```

Comment:
Related to my comment, should we add a test for:

```
# app -(br)-> gtest/0.1 -(br)-> cmake/0.1 -(br)->gtest/0.1 ....
                       \ (br)-> foo/0.1  --(br) / (edited) 
```

Reply:
Trying to add this structure, but that can never happen. build-requires do not close the diamond, because they are not visible downstream. And in any case, only the first branch is expanded first, causing exactly the same scenario and resulting graph than the current existing test.